File: C++/sliding-window/121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int maxP = 0;

        int l = 0, r = 1;
        while (r < prices.size())
        {
            if (prices[l] < prices[r])
            {
                maxP = max(maxP, prices[r] - prices[l]);
                r++;
            }
            else
            {
                l = r;
                r++;
            }
        }
    }
};
// @lc code=end

File: Rust/sliding-window/121.best-time-to-buy-and-sell-stock.rs
/*
 * @lc app=leetcode id=121 lang=rust
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_p: i32 = 0;
        let mut l: i32 = 0;
        let mut r: i32 = 1;

        while r < prices.len() {
            if prices[l] < prices[r] {
                max_p = max_p.max(prices[r] - prices[l]);
            } else {
                l = r;
            }

            r += 1;
        }

        return max_p;
    }
}
// @lc code=end
